crates/bevy_math/src/primitives/dim3.rs

```diff
@@ -629,6 +629,166 @@ impl Torus {
     }
 }
 
+/// A 3D triangle primitive.
+#[derive(Clone, Copy, Debug, PartialEq)]
+#[cfg_attr(feature = "serialize", derive(serde::Serialize, serde::Deserialize))]
+pub struct Triangle3d {
+    /// The vertices of the triangle
+    pub vertices: [Vec3; 3],
+}
+
+/// An error that occurs when a triangle is degenerate, meaning it has zero area.
+/// This can happen when the three vertices are collinear or nearly collinear.
+#[derive(Debug, Clone)]
+pub struct DegenerateTriangleError;
+
+impl std::fmt::Display for DegenerateTriangleError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Triangle is degenerate")
+    }
+}
+
+impl Primitive3d for Triangle3d {}
+
+impl Default for Triangle3d {
+    /// Returns the default [`Triangle3d`] with vertices at the `[-0.5, 0.0, 0.0]`, `[0.5, 0.0, 0.0]`, and `[0.0, 0.5, 0.0]`
+    fn default() -> Self {
+        Self {
+            vertices: [
+                Vec3::new(-0.5, 0.0, 0.0),
+                Vec3::new(0.5, 0.0, 0.0),
+                Vec3::new(0.0, 0.5, 0.0),
+            ],
+        }
+    }
+}
+
+impl Triangle3d {
+    /// Create a new [`Triangle3d`] from points `a`, `b`, and `c`
+    #[inline(always)]
+    pub fn new(a: Vec3, b: Vec3, c: Vec3) -> Self {
+        Self {
+            vertices: [a, b, c],
+        }
+    }
+
+    /// Get the area of the triangle
+    #[inline(always)]
+    pub fn area(&self) -> f32 {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        ab.cross(ac).length() / 2.0
+    }
+
+    /// Get the perimeter of the triangle
+    #[inline(always)]
+    pub fn perimeter(&self) -> f32 {
+        let [a, b, c] = self.vertices;
+        let ab = a.distance(b);
+        let bc = b.distance(c);
+        let ca = c.distance(a);
+        ab + bc + ca
+    }
+
+    /// Get the normal of the triangle in the direction of the right-hand rule, assuming
+    /// the vertices are ordered in a counter-clockwise direction.
+    ///
+    /// The normal is computed as the cross product of the vectors `ab` and `ac`.
+    ///
+    /// # Errors
+    ///
+    /// Returns [`DegenerateTriangleError`] if the triangle is degenerate, meaning it has zero area.
+    #[inline(always)]
+    pub fn normal(&self) -> Result<Dir3, DegenerateTriangleError> {
+        let [a, b, c] = self.vertices;
+        let ab = b - a;
+        let ac = c - a;
+        let cross_product = ab.cross(ac);
+        if cross_product.length() < f32::EPSILON {
+            return Err(DegenerateTriangleError);
+        }
+        Ok(Dir3::new_unchecked(cross_product.normalize()))
+    }
```

Comment:
This is now computing the length twice, which is an extra square root. The docs should probably also mention the NaN case.

The implementation should be as simple as this:

```suggestion
    pub fn normal(&self) -> Result<Dir3, DegenerateTriangleError> {
        let [a, b, c] = self.vertices;
        let ab = b - a;
        let ac = c - a;
        Dir3::new(ab.cross(ac)).map_err(|_| DegenerateTriangleError)
    }
```

I'm not entirely sure if we need the `DegenerateTriangleError` either. We could just return `InvalidDirectionError` here (returned by `Dir3::new`), which also gives more data about the error. Ultimately the error is that a valid normal direction could not be computed because of the triangle being degenerate, and imo `InvalidDirectionError` would work for this too.